File: src/android/extensionNotifications.ts
import {
  NotificationError,
  checkNotificationId,
  mergeNotificationOptions,
  normalizeNotificationOptions,
  toNativeNotification,
  type NotificationOptions
} from '@core/extensions/api/notifications'
import type { AttachedExtension } from './extensionApi'

/**
 * `chrome.notifications` on Android: Chrome's option rules from the shared module, the system
 * notification from Kotlin (`ext/ExtensionNotifications.kt`) on a channel per extension, named
 * after it so the user finds it under the app's notification settings. `update` re-posts the
 * merged notification under the same tag, which Android edits in place; a tap, a button or a
 * swipe comes back as `ext.notification` and is `onClicked`, `onButtonClicked`, `onClosed`.
 * A tap or a button dismisses the card (as it does in Chrome), so `onClosed(id, true)` follows.
 * `getPermissionLevel` is what the app's own notification permission says.
 */

/** One notification as Kotlin shows it: Chrome's templates flattened, plus what Android can render richer. */
export interface ShownNotification {
  notificationId: string
  /** The extension's name (its channel's name in the system settings). */
  extensionName: string
  title: string
  body: string
  /** `contextMessage`: the sub text line. */
  subText: string
  /** `iconUrl` as given: a `data:` URL or a path inside the extension (Kotlin resolves it). */
  iconUrl: string | null
  /** The hero picture of an `image` notification, resolved like the icon. */
  imageUrl: string | null
  /** 0 to 100 for a `progress` notification, else null. */
  progress: number | null
  buttons: string[]
  silent: boolean
  /** Chrome's `priority` -2..2 (Android maps it onto its pre-channel priorities). */
  priority: number
  /** `eventTime` in ms since the epoch, or null for now. */
  eventTime: number | null
}

export interface NotificationsHost {
  /** Show one notification (or replace the one under the same id in place). */
  show(extensionId: string, notification: ShownNotification): void
  /** Take one down without an `onClosed` (`clear`, a replacement, the extension going). */
  hide(extensionId: string, notificationId: string): void
  /** The extension is gone for good: its channel goes with its notifications. */
  forget(extensionId: string): void
  /** Whether the app may post notifications right now (`getPermissionLevel`). */
  allowed(): Promise<boolean>
  emit(extensionId: string, ns: string, name: string, args: unknown[]): void
}

/** What Kotlin reports about a shown notification (`ext.notification`). */
export interface NotificationEvent {
  extensionId: string
  notificationId: string
  event: 'clicked' | 'button' | 'closed'
  /** The button's index for `button`. */
  index?: number
}

export class AndroidNotifications {
  /** Extension id → notification id → its effective options. */
  private readonly shown = new Map<string, Map<string, NotificationOptions>>()
  private seq = 0

  constructor(private readonly host: NotificationsHost) {}

  call(ext: AttachedExtension, method: string, args: unknown[]): unknown {
    switch (method) {
      case 'create':
        return this.create(ext, args[0], args[1])
      case 'update':
        return this.update(ext, args[0], args[1])
      case 'clear':
        return this.clear(ext, args[0])
      case 'getAll': {
        const out: Record<string, true> = {}
        for (const id of this.shown.get(ext.record.id)?.keys() ?? []) out[id] = true
        return out
      }
      case 'getPermissionLevel':
        return this.host.allowed().then((allowed) => (allowed ? 'granted' : 'denied'))
    }
    throw new Error(`chrome.notifications.${method} is not implemented on Zenium for Android`)
  }

  /** The extension is going away (uninstall): its notifications and channel with it. */
  forget(extensionId: string): void {
    this.shown.delete(extensionId)
    this.host.forget(extensionId)
  }

  /** A tap, a button or a swipe on a shown notification, from Kotlin. */
  onEvent(e: NotificationEvent): void {
    const { extensionId, notificationId } = e
    // A notification may outlive the process (it stays in the shade); the events still count.
    this.shown.get(extensionId)?.delete(notificationId)
    if (e.event === 'clicked') {
      this.host.emit(extensionId, 'notifications', 'onClicked', [notificationId])
    } else if (e.event === 'button') {
      this.host.emit(extensionId, 'notifications', 'onButtonClicked', [
        notificationId,
        typeof e.index === 'number' ? e.index : 0
      ])
    }
    // Android dismisses the card on a tap (autoCancel) and Kotlin cancels it after a button; a
    // swipe reports `closed` directly. Each ends with Chrome's `onClosed(id, byUser)`.
    this.host.emit(extensionId, 'notifications', 'onClosed', [notificationId, true])
  }

  /**
   * `create(notificationId?, options)`: the shim aligns the optional id to its slot (`undefined`
   * when omitted), but a bare `(options)` call is accepted as well. A second `create` under an
   * id in use replaces the notification, silently.
   */
  private create(ext: AttachedExtension, first: unknown, second: unknown): string {
    const explicit = typeof first === 'string' ? first : null
    const raw = explicit !== null || first === undefined || first === null ? second : first
    const id = explicit || this.newId()
    try {
      checkNotificationId(id)
      const options = normalizeNotificationOptions(raw, true)
      this.listFor(ext.record.id).set(id, options)
      this.host.show(ext.record.id, this.spec(ext, id, options))
      return id
    } catch (error) {
      throw toError(error)
    }
  }

  private update(ext: AttachedExtension, id: unknown, raw: unknown): boolean {
    if (typeof id !== 'string') throw new Error('Invalid notificationId')
    const list = this.shown.get(ext.record.id)
    const current = list?.get(id)
    if (!list || !current) return false
    try {
      const patch = normalizeNotificationOptions(raw, false)
      const merged = mergeNotificationOptions(current, patch)
      list.set(id, merged)
      this.host.show(ext.record.id, this.spec(ext, id, merged))
      return true
    } catch (error) {
      throw toError(error)
    }
  }

  private clear(ext: AttachedExtension, id: unknown): boolean {
    if (typeof id !== 'string') throw new Error('Invalid notificationId')
    const list = this.shown.get(ext.record.id)
    if (!list?.has(id)) return false
    list.delete(id)
    this.host.hide(ext.record.id, id)
    this.host.emit(ext.record.id, 'notifications', 'onClosed', [id, false])
    return true
  }

  private listFor(extensionId: string): Map<string, NotificationOptions> {
    let list = this.shown.get(extensionId)
    if (!list) {
      list = new Map()
      this.shown.set(extensionId, list)
    }
    return list
  }

  private spec(
    ext: AttachedExtension,
    id: string,
    options: NotificationOptions
  ): ShownNotification {
    // With the sub text line taking `contextMessage`, the body keeps the message, list items and
    // the progress percentage (the bar shows it too).
    const flat = toNativeNotification(options, true)
    return {
      notificationId: id,
      extensionName: ext.manifest.name,
      title: flat.title,
      body: flat.body,
      subText: flat.subtitle,
      iconUrl: flat.iconUrl,
      imageUrl: options.type === 'image' ? (options.imageUrl ?? null) : null,
      progress:
        options.type === 'progress' && options.progress !== undefined
          ? Math.round(Math.min(100, Math.max(0, options.progress)))
          : null,
      buttons: flat.buttons,
      silent: flat.silent,
      priority: options.priority ?? 0,
      eventTime: options.eventTime ?? null
    }
  }

  /** Chrome hands out a random token when the extension names no id. */
  private newId(): string {
    const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_'
    let out = ''
    for (let i = 0; i < 20; i++) out += alphabet[Math.floor(Math.random() * alphabet.length)]
    this.seq += 1
    return `${out}${this.seq.toString(36)}`
  }
}

/** The event payload Kotlin posts, checked; null when it is not one. */
export function notificationEvent(raw: unknown): NotificationEvent | null {
  if (typeof raw !== 'object' || raw === null) return null
  const p = raw as Record<string, unknown>
  if (typeof p.id !== 'string' || typeof p.notificationId !== 'string') return null
  if (p.event !== 'clicked' && p.event !== 'button' && p.event !== 'closed') return null
  const out: NotificationEvent = {
    extensionId: p.id,
    notificationId: p.notificationId,
    event: p.event
  }
  if (typeof p.index === 'number' && Number.isInteger(p.index) && p.index >= 0) out.index = p.index
  return out
}

function toError(error: unknown): Error {
  if (error instanceof NotificationError) return new Error(error.message)
  return error instanceof Error ? error : new Error(String(error))
}

File: android/app/src/main/kotlin/app/zen/chromium/ext/ExtensionNotifications.kt
package app.zen.chromium.ext

import android.Manifest
import android.app.NotificationChannel
import android.app.NotificationChannelGroup
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Build
import android.os.Handler
import android.os.Looper
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import app.zen.chromium.MainActivity
import app.zen.chromium.R
import org.json.JSONObject
import java.io.File
import java.util.concurrent.Executor

/**
 * `chrome.notifications` on the system shade: one notification channel per extension, named
 * after it (the user finds it under the app's notification settings, in an "Extensions" group)
 * and one card per notification id, edited in place by a re-post under the same tag. Chrome's
 * templates arrive flattened from the core (`src/android/extensionNotifications.ts`): title,
 * body, sub text, up to two buttons, a progress value, the icon and the hero picture as a
 * `data:` URL or a path inside the extension's directory.
 *
 * A tap and a button go through [MainActivity] (the extension's handler mostly opens a tab, which
 * wants the app in front), a swipe through a broadcast; each reaches [onEvent] as
 * `clicked` / `button` / `closed`. Android keeps the card across a process death, so a tap may
 * arrive before the extension runs again; [Extensions] holds it until then.
 */
class ExtensionNotifications(
    private val context: Context,
    private val io: Executor,
    private val onEvent: (extensionId: String, notificationId: String, event: String, index: Int) -> Unit
) {
    private val manager = NotificationManagerCompat.from(context)
    private val main = Handler(Looper.getMainLooper())
    /** Channels created in this process (creating one again is a no-op, but a lookup is cheaper). */
    private val channels = HashSet<String>()
    /** Per extension, the notification ids shown, to take them all down at once. */
    private val shown = HashMap<String, MutableSet<String>>()

    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(c: Context, intent: Intent) {
            onIntent(intent)
        }
    }

    init {
        ContextCompat.registerReceiver(context, receiver, IntentFilter(ACTION_DISMISSED), ContextCompat.RECEIVER_NOT_EXPORTED)
    }

    /** Whether the app may post notifications right now (`getPermissionLevel`). */
    fun allowed(): Boolean =
        Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU ||
            ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED

    /**
     * Show (or replace) one notification. `dir` is the extension's directory for relative icon
     * paths; the images are decoded off the main thread and the card posted from there.
     */
    fun show(extensionId: String, dir: File?, n: JSONObject) {
        val notificationId = n.optString("notificationId")
        if (notificationId.isEmpty()) return
        val extensionName = n.optString("extensionName").ifEmpty { "Extension" }
        ensureChannel(extensionId, extensionName)
        shown.getOrPut(extensionId) { LinkedHashSet() }.add(notificationId)
        val iconUrl = n.optString("iconUrl").ifEmpty { null }
        val imageUrl = n.optString("imageUrl").ifEmpty { null }
        io.execute {
            val icon = iconUrl?.let { loadImage(dir, it, MAX_ICON_PX) }
            val picture = imageUrl?.let { loadImage(dir, it, MAX_PICTURE_PX) }
            main.post {
                // Taken down in the meantime: nothing to post.
                if (shown[extensionId]?.contains(notificationId) == true) post(extensionId, notificationId, n, icon, picture)
            }
        }
    }

    /** Take one down without an event (`clear`, the extension going). */
    fun hide(extensionId: String, notificationId: String) {
        shown[extensionId]?.remove(notificationId)
        manager.cancel(tag(extensionId, notificationId), NOTIFICATION_ID)
    }

    /** The extension is gone: its cards and its channel (Android remembers the channel's settings). */
    fun forget(extensionId: String) {
        for (id in shown.remove(extensionId).orEmpty()) manager.cancel(tag(extensionId, id), NOTIFICATION_ID)
        if (channels.remove(channelId(extensionId))) {
            runCatching { systemManager().deleteNotificationChannel(channelId(extensionId)) }
        }
    }

    /** A tap, button or swipe intent (from [MainActivity] or the dismiss receiver). */
    fun onIntent(intent: Intent): Boolean {
        val extensionId = intent.getStringExtra(EXTRA_EXTENSION) ?: return false
        val notificationId = intent.getStringExtra(EXTRA_NOTIFICATION) ?: return false
        val event = intent.getStringExtra(EXTRA_EVENT) ?: return false
        val index = intent.getIntExtra(EXTRA_INDEX, 0)
        if (event != EVENT_CLOSED) {
            // Android dismisses a tapped card itself (autoCancel); a button leaves it, Chrome does not.
            manager.cancel(tag(extensionId, notificationId), NOTIFICATION_ID)
        }
        shown[extensionId]?.remove(notificationId)
        onEvent(extensionId, notificationId, event, index)
        return true
    }

    fun destroy() {
        runCatching { context.unregisterReceiver(receiver) }
    }

    private fun post(extensionId: String, notificationId: String, n: JSONObject, icon: Bitmap?, picture: Bitmap?) {
        if (!allowed()) return
        val title = n.optString("title")
        val body = n.optString("body")
        val subText = n.optString("subText")
        val builder = NotificationCompat.Builder(context, channelId(extensionId))
            .setSmallIcon(R.drawable.ic_stat_zenium)
            .setContentTitle(title)
            .setContentText(body.lineSequence().firstOrNull() ?: "")
            .setAutoCancel(true)
            .setOnlyAlertOnce(true)
            .setPriority(priority(n.optInt("priority", 0)))
            .setCategory(NotificationCompat.CATEGORY_MESSAGE)
            .setContentIntent(activityIntent(extensionId, notificationId, EVENT_CLICKED, 0))
            .setDeleteIntent(dismissIntent(extensionId, notificationId))
        if (subText.isNotEmpty()) builder.setSubText(subText)
        if (n.optBoolean("silent", false)) builder.setSilent(true)
        if (icon != null) builder.setLargeIcon(icon)
        val eventTime = n.optLong("eventTime", 0L)
        if (eventTime > 0) builder.setWhen(eventTime).setShowWhen(true)
        val progress = if (n.isNull("progress")) -1 else n.optInt("progress", -1)
        if (progress in 0..100) builder.setProgress(100, progress, false)
        if (picture != null) {
            builder.setStyle(NotificationCompat.BigPictureStyle().bigPicture(picture).setSummaryText(body))
        } else if (body.contains('\n') || body.length > 40) {
            builder.setStyle(NotificationCompat.BigTextStyle().bigText(body))
        }
        val buttons = n.optJSONArray("buttons")
        if (buttons != null) {
            for (i in 0 until minOf(buttons.length(), MAX_BUTTONS)) {
                val label = buttons.optString(i)
                if (label.isEmpty()) continue
                builder.addAction(0, label, activityIntent(extensionId, notificationId, EVENT_BUTTON, i))
            }
        }
        runCatching { manager.notify(tag(extensionId, notificationId), NOTIFICATION_ID, builder.build()) }
    }

    private fun ensureChannel(extensionId: String, extensionName: String) {
        val id = channelId(extensionId)
        if (!channels.add(id)) return
        val system = systemManager()
        system.createNotificationChannelGroup(NotificationChannelGroup(GROUP_ID, GROUP_NAME))
        system.createNotificationChannel(
            NotificationChannel(id, extensionName, NotificationManager.IMPORTANCE_DEFAULT).apply {
                description = "Notifications from the $extensionName extension"
                group = GROUP_ID
            }
        )
    }

    private fun systemManager(): NotificationManager =
        context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

    private fun activityIntent(extensionId: String, notificationId: String, event: String, index: Int): PendingIntent {
        val intent = Intent(context, MainActivity::class.java)
            .setAction(ACTION_OPENED)
            .addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
            .putExtra(EXTRA_EXTENSION, extensionId)
            .putExtra(EXTRA_NOTIFICATION, notificationId)
            .putExtra(EXTRA_EVENT, event)
            .putExtra(EXTRA_INDEX, index)
        return PendingIntent.getActivity(
            context, requestCode(extensionId, notificationId, event, index), intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
    }

    private fun dismissIntent(extensionId: String, notificationId: String): PendingIntent {
        val intent = Intent(ACTION_DISMISSED).setPackage(context.packageName)
            .putExtra(EXTRA_EXTENSION, extensionId)
            .putExtra(EXTRA_NOTIFICATION, notificationId)
            .putExtra(EXTRA_EVENT, EVENT_CLOSED)
        return PendingIntent.getBroadcast(
            context, requestCode(extensionId, notificationId, EVENT_CLOSED, 0), intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
    }

    /** One pending intent per (card, event, button): the extras differ, so the codes have to. */
    private fun requestCode(extensionId: String, notificationId: String, event: String, index: Int): Int =
        "$extensionId/$notificationId/$event/$index".hashCode()

    private fun loadImage(dir: File?, url: String, maxPx: Int): Bitmap? {
        val bytes = imageBytes(dir, url) ?: return null
        return runCatching {
            val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size, bounds)
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size, BitmapFactory.Options().apply { inSampleSize = sampleSize(bounds.outWidth, bounds.outHeight, maxPx) })
        }.getOrNull()
    }

    companion object {
        /** The activity intent of a tap or a button (`MainActivity.handleIntent`). */
        const val ACTION_OPENED = "app.zen.chromium.EXT_NOTIFICATION"
        /** The broadcast of a swipe (the receiver above). */
        const val ACTION_DISMISSED = "app.zen.chromium.EXT_NOTIFICATION_DISMISSED"
        const val EXTRA_EXTENSION = "ext"
        const val EXTRA_NOTIFICATION = "notification"
        const val EXTRA_EVENT = "event"
        const val EXTRA_INDEX = "index"
        const val EVENT_CLICKED = "clicked"
        const val EVENT_BUTTON = "button"
        const val EVENT_CLOSED = "closed"
        const val GROUP_ID = "zenium.extensions"
        const val GROUP_NAME = "Extensions"
        /** Every card has its own tag; the numeric id is constant. */
        const val NOTIFICATION_ID = 1
        const val MAX_BUTTONS = 2
        private const val MAX_ICON_PX = 256
        private const val MAX_PICTURE_PX = 1024

        fun channelId(extensionId: String): String = "zenium.ext.$extensionId"
        fun tag(extensionId: String, notificationId: String): String = "zenium.ext/$extensionId/$notificationId"

        /** Chrome's -2..2 onto Android's pre-channel priorities (channels decide on Android 8+). */
        fun priority(chrome: Int): Int = when {
            chrome >= 2 -> NotificationCompat.PRIORITY_MAX
            chrome == 1 -> NotificationCompat.PRIORITY_HIGH
            chrome == -1 -> NotificationCompat.PRIORITY_LOW
            chrome <= -2 -> NotificationCompat.PRIORITY_MIN
            else -> NotificationCompat.PRIORITY_DEFAULT
        }

        /**
         * The bytes of an `iconUrl` / `imageUrl` as Chrome resolves them: a `data:` URL, or a path
         * inside the extension (`chrome-extension://<id>/x.png`, `/x.png` and `x.png` alike; one
         * that escapes the directory is nothing). Anything else (an http URL) is not fetched;
         * Chrome refuses those too.
         */
        fun imageBytes(dir: File?, url: String): ByteArray? = runCatching {
            if (url.startsWith("data:")) {
                val comma = url.indexOf(',')
                if (comma < 0) return null
                return if (url.substring(0, comma).endsWith(";base64")) java.util.Base64.getMimeDecoder().decode(url.substring(comma + 1).trim())
                else java.net.URLDecoder.decode(url.substring(comma + 1), "UTF-8").toByteArray()
            }
            val relative = url.replace(Regex("^chrome-extension://[a-p]{32}"), "")
            if (Regex("^[a-zA-Z][a-zA-Z0-9+.-]*:").containsMatchIn(relative)) return null
            val file = fileIn(dir ?: return null, relative) ?: return null
            if (file.isFile) file.readBytes() else null
        }.getOrNull()

        /** The power-of-two subsampling that brings a `width` x `height` image near `maxPx`. */
        fun sampleSize(width: Int, height: Int, maxPx: Int): Int {
            var sample = 1
            while (width / sample > maxPx * 2 || height / sample > maxPx * 2) sample *= 2
            return sample
        }

        private fun fileIn(dir: File, path: String): File? {
            val file = File(dir, path.trimStart('/'))
            val canonical = runCatching { file.canonicalPath }.getOrNull() ?: return null
            val root = runCatching { dir.canonicalPath }.getOrNull() ?: return null
            if (!canonical.startsWith(root + File.separator)) return null
            return file
        }
    }
}
